### PR TITLE
add spot instance ui stuff

### PR DESCRIPTION
**Pull request by ubergesundheit**
Towards giantswarm/giantswarm#8347

Related to https://github.com/giantswarm/api/pull/999

### CODE REVIEW

**Comment by marians** [comment]
@ubergesundheit Could you please post some screenshots here?

**Comment by marians** [comment]
Some changes:

- second number picker should be "Spot instance percentage". Logic would be `on_demand_percentage = 100 - user-defined_value`.
- Font weight of labels should not be bold.
- New attributes (alike instance types, base on-demand capacity, spot percentage) should be active only from the new release on.

**Comment by ubergesundheit** [comment]
Screen Shot 2020-03-27 at 11 18 23

**Comment by ubergesundheit** [comment]
image

**Comment by ubergesundheit** [comment]
@giantswarm/sig-ux @giantswarm/chapter-front-end this is ready for review

File: src/components/Cluster/ClusterDetail/AddNodePool.js [hunk 1]
```diff
@@ -177,20 +203,46 @@ class AddNodePool extends Component {
...
     awsInstanceTypes: {},
+    allowSpotInstances: false,
+    allowAlikeInstances: false,
   };
 
   componentDidMount() {
     this.setState({
       awsInstanceTypes: JSON.parse(window.config.awsCapabilitiesJSON),
+      ...this.computeInstanceCapabilities(),
     });
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/AddNodePool.js]: Do these have to be in state? Do they ever change?

File: src/components/Cluster/ClusterDetail/AddNodePool.js [hunk 2]
```diff
@@ -352,8 +446,48 @@ class AddNodePool extends Component {
...
+              <FlexWrapperDiv className='alike-wrapper'>
+                <CheckboxWrapper>
+                  <input
+                    className='original-checkbox'
+                    type='checkbox'
+                    checked={this.state.aws.useAlike}
+                    onChange={this.setUseAlikeInstancesEnabled}
+                    id='aws-alike-instances-enabled'
+                  />
+                </CheckboxWrapper>
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/AddNodePool.js]: Can this be converted to a generic UI component? We could use a styled checkbox 😬

File: src/components/Cluster/ClusterDetail/NodePool.js [hunk 3]
```diff
@@ -119,7 +121,27 @@ class NodePool extends Component {
...
+              {node_spec.aws.use_alike_instance_types
+                ? 'mixed'
+                : node_spec.aws.instance_type}
+            </Code>
+            <NodesWrapper>
+              {instance_distribution &&
+              typeof instance_distribution.on_demand_base_capacity !==
+                'undefined'
+                ? instance_distribution.on_demand_base_capacity
+                : '-'}
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/NodePool.js]:
```suggestion
              {instance_distribution?.on_demand_base_capacity ?? '-'}
```

File: src/components/Cluster/ClusterDetail/NodePool.js [hunk 4]
```diff
@@ -119,7 +121,27 @@ class NodePool extends Component {
...
+              /
+              {instance_distribution &&
+              typeof instance_distribution.on_demand_percentage_above_base_capacity !==
+                'undefined'
+                ? `${
+                    100 /* eslint-disable-line no-magic-numbers */ -
+                    instance_distribution.on_demand_percentage_above_base_capacity
+                  } %`
+                : '-'}
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/NodePool.js]: Can you move this to a separate function, please?

File: src/components/UI/NumberPicker.js [hunk 5]
```diff
@@ -171,7 +171,7 @@ class NumberPicker extends React.Component {
       {
         inputValue: value,
         value,
-        valid: value ? true : false,
+        valid: value || (min <= 0 && value === 0) ? true : false,
```
Comment (axbarsan) [inline src/components/UI/NumberPicker.js]:
```suggestion
        valid: Boolean(value) || (min <= 0 && value === 0),
```

**Comment by marians** [comment]
I'd like to test this but on ginger the newest release is 11.0.1. Can you make that testable please?

File: src/components/Cluster/ClusterDetail/AddNodePool.js (hunk 2 above)
Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/AddNodePool.js]: Sure! I can do it in another PR if you don't mind?

File: src/components/Cluster/ClusterDetail/NodePool.js (hunk 3 above)
Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/NodePool.js]: I had this previously like this:
`instance_distribution?.on_demand_base_capacity || '-'`
This broke for value `0`. I'll test this though

File: src/components/Cluster/ClusterDetail/AddNodePool.js (hunk 1 above)
Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/AddNodePool.js]: On cluster creation, when the user selects another version.

File: src/components/Cluster/ClusterDetail/NodePool.js (hunk 3 above)
Comment (axbarsan) [reply src/components/Cluster/ClusterDetail/NodePool.js]: `||` will guard against any empty value (including `0`),  but `??` will only guard against nullable values (`undefined` or `null`)

**Comment by ubergesundheit** [comment]
@marians https://github.com/giantswarm/happa/pull/1270/commits/6acfcc2b4490fc39ab6183fe3d1ad95e5654b957 is now deployed do ginger

**Comment by marians** [comment]
There is some style glitch at the checkbox on Mac OS with Chrome v80.

image

**Comment by marians** [comment]
Please align the `Allow usage of similar instance types` on the left-hand axis under the instance type drop-down.

Also please make the font size 16px instead of 14px.

Alignment mock-up (with wrong font size):

image

**Comment by marians** [comment]
### Label width, font weight, and line height

Currently

image

Desired

image

```css
font-weight: 300;
font-size: 16px;
line-height: 32px;
```

**Comment by marians** [comment]
These error message appear in an odd position

image

**Comment by marians** [comment]
One last thing: How about moving the `Instance distribution` section directly above the `Scaling range` section? Both are somehow related to the node pool size and users might want to adjust the base on-demand capacity according to the scaling range they decide for.

**Comment by axbarsan** [comment]
A quick issue: If you have multiple node pools created, checking the `Allow usage of similar instance types`, for any of the node pools, will check the checkbox on the first node pool (probably because they all have the same id)

File: src/components/Cluster/ClusterDetail/AddNodePool.js [hunk 6]
```diff
@@ -352,8 +446,48 @@ class AddNodePool extends Component {
...
+                  <p>Allow usage of similar instance types</p>
+                </label>
+              </FlexWrapperDiv>
+            )}
           </FlexWrapperDiv>
         </label>
+        {this.state.allowSpotInstances && (
+          <SpotValuesLabel>
+            <span className='label-span'>Instance distribution</span>
+            <NumberPicker
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/AddNodePool.js]: This has no maximum value, and if I try to use a large enough number, the node pool creation fails

File: src/components/Cluster/ClusterDetail/NodePool.js (hunk 4 above)
Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/NodePool.js]: check out 2a3942e

File: src/components/Cluster/ClusterDetail/NodePool.js [hunk 7]
```diff
@@ -93,6 +93,22 @@ class NodePool extends Component {
...
+
+    const baseCapacity = instance_distribution?.on_demand_base_capacity ?? '-';
+    let spotInstancePercentage =
+      instance_distribution?.on_demand_percentage_above_base_capacity ?? '-';
+
+    /* eslint-disable-next-line no-magic-numbers */
+    spotInstancePercentage =
+      spotInstancePercentage !== '-'
+        ? `${100 - spotInstancePercentage} %`
+        : '-';
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/NodePool.js]: Apologize for the nitpicks 😅 

```suggestion
if (spotInstancePercentage !== '-') {
  spotInstancePercentage = `${100 - spotInstancePercentage} %`;
}
```

File: src/components/Cluster/ClusterDetail/AddNodePool.js (hunk 6 above)
Comment (teemow) [reply src/components/Cluster/ClusterDetail/AddNodePool.js]: how large?

@ubergesundheit we can probably limit this to 1000 for now.

Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/AddNodePool.js]: so above `32767`?

Comment (axbarsan) [reply src/components/Cluster/ClusterDetail/AddNodePool.js]: I tried `999999999999999999999999999999` 😬

**Comment by ubergesundheit** [comment]
I hope I adressed all of the issues. One thing though: Selecting suitable icons from fortawesome for the checkbox states. I'll do that once I figure out how to use lastpass

File: src/components/UI/Checkbox.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import React from 'react';
+
+const CheckboxWrapper = styled.div``;
+
+const CheckboxLabel = styled.span`
+  margin-left: 5px;
+`;
+
+/* eslint-disable-next-line no-empty-function */
+const Checkbox = ({ checked, label, onChange = () => {} }) => {
```
Comment (axbarsan) [inline src/components/UI/Checkbox.js]: A prop's default value is usually set by using the `defaultProps` static property of a component.

https://reactjs.org/docs/typechecking-with-proptypes.html#default-prop-values

File: src/components/UI/Checkbox.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  return (
+    <CheckboxWrapper onClick={() => onChange(!checked)}>
+      <i
+        className={`fa fa-${checked ? 'check-box' : 'check-box-outline-blank'}`}
+      />
+      <CheckboxLabel className='checkbox-label'>{label}</CheckboxLabel>
+    </CheckboxWrapper>
+  );
```
Comment (axbarsan) [inline src/components/UI/Checkbox.js]: This approach is not accessible, nor compatible with the native checkbox.
Custom appearances for checkboxes can be created with just CSS.

We can work together on that in a separate PR, as we should move on with this one.

**Comment by ubergesundheit** [comment]
As discussed in sig-ux sync, I've moved the spot instance number inputs behind a checkbox

image

**Comment by teemow** [comment]
Thanks @ubergesundheit. But there is no checkbox anymore. You can't tick the feature on/off with a checkbox. It's fine if you want to keep this in the UI. But you need to make sure that the form contains 0 spot instance percentage once submitted to the api.

**Comment by ubergesundheit** [comment]
TODOs for now:

- [ ] Change the version gate to version `11.2.0`

Node pools table:

- [ ] Show mixed instance types (For example tooltip) 
- [ ] Single column for showing the number of spot instances in node pool (name: Spot)
- [ ] Show spot instance user configuration in tooltip of spot instance number column

Node pool creation dialog:

- [ ] Remove `AWS` from `Enable AWS Spot Instances` from creation dialog checkbox
- [ ] Add more bottom margin for labels under Instance distribution help text
- [ ] Add units `percent` & `instances` behind number input

File: __tests__/AccountSettings.js [hunk 10]
```diff
@@ -1,13 +1,13 @@
 import { getInstallationInfo } from 'model/services/giantSwarm';
+import { getConfiguration } from 'model/services/metadata';
```
Comment (ubergesundheit) [inline __tests__/AccountSettings.js:2]: I don't know how the changes in these test files got in here, I cannot remember doing them. For now I suggest to ignore them. Or should I revert these?

Comment (axbarsan) [reply __tests__/AccountSettings.js:2]: They're part of the version check. They are in master, probably appeared here after the merge

Comment (ubergesundheit) [reply __tests__/AccountSettings.js:2]: ok :+1:

File: src/components/Cluster/ClusterDetail/NodePool.tsx [hunk 11]
```diff
@@ -35,12 +38,23 @@ const NPViewAndEditNameStyled = styled<
...
   height: 30px;
   line-height: 31px;
   text-align: center;
   border-radius: 3px;
+  white-space: nowrap;
+  ${({ highlight }: INodesWrapperProps) =>
+    highlight &&
+    css`
+      background-color: ${theme.colors.goldBackground};
+    `}
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/NodePool.tsx:57]: This should also work:
```suggestion
  background-color: ${({ theme, highlight }) => highlight && theme.colors.goldBackground};
```

File: src/components/Cluster/ClusterDetail/NodePool.tsx [hunk 12]
```diff
@@ -210,14 +288,21 @@ class NodePool extends Component<INodePoolsProps, INodePoolsState> {
...
-            <NodesWrapper
-              style={{
-                background:
-                  current < desired ? theme.colors.goldBackground : undefined,
-              }}
-            >
-              {current}
-            </NodesWrapper>
+            <NodesWrapper highlight={current < desired}>{current}</NodesWrapper>
+            {typeof spot_instances !== undefined || spot_instances !== null ? (
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/NodePool.tsx:292]: This can be just this, right?
```suggestion
            {spot_instances ? (
```

Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/NodePool.tsx:292]: does not work for `spot_instances === 0`

Comment (axbarsan) [reply src/components/Cluster/ClusterDetail/NodePool.tsx:292]: Oh ok, did not check the data type, thought it was an object 👍

File: src/components/Cluster/ClusterDetail/NodePool.tsx (hunk 11 above)
Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/NodePool.tsx:57]: Whats the type of the parameter of this function then (`{ theme, highlight }`)?

Comment (axbarsan) [reply src/components/Cluster/ClusterDetail/NodePool.tsx:57]: Probably something like `INodesWrapperProps | undefined`, but I remember that we decided to not type styled components

It's also written here, under the `Style` section
https://github.com/giantswarm/happa/pull/1268

File: src/components/Cluster/ClusterDetail/NodePool.tsx [hunk 13]
```diff
@@ -35,12 +38,23 @@ const NPViewAndEditNameStyled = styled<
   }
 `;
 
+interface INodesWrapperProps {
+  highlight?: boolean;
+  children: React.ReactNode;
+}
+
 const NodesWrapper = styled.div`
```
Comment (axbarsan) [inline src/components/Cluster/ClusterDetail/NodePool.tsx:46]: I think I know what you meant
```suggestion
const NodesWrapper = styled<{highlight?: boolean}>.div`
```

This should do it

**Comment by ubergesundheit** [comment]
Changes since yesterday:

- Only display `+N` for alike instances when > 1
- Use lighter color for nodepool with enabled alike instances
- Tooltip of nodepool with enabled alike instances now reads `Alike instances enabled. Currently used: ...`
- Instance type of nodepools with and without alike instances are now vertically aligned

File: src/components/Cluster/ClusterDetail/NodePool.tsx (hunk 13 above)
Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/NodePool.tsx]: thanks :pray:

File: src/components/Cluster/ClusterDetail/NodePool.tsx (hunk 12 above)
Comment (ubergesundheit) [reply src/components/Cluster/ClusterDetail/NodePool.tsx:290]: tuns out I am a little dummy boy :) this should've been `typeof spot_instances !== 'undefined' && spot_instances !== null`. But I'll change it to `typeof spot_instances === 'number'`

**Comment by marians** [comment]
Thanks a lot @ubergesundheit ! 💐

**Pull request merged**